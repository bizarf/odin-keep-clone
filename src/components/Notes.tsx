import React from "react";
import {
    MdOutlineNotificationAdd,
    MdOutlinePersonAddAlt,
    MdOutlinePalette,
    MdOutlineImage,
    MdOutlineArchive,
    MdOutlineMoreVert,
    MdOutlinePushPin,
    MdPushPin,
} from "react-icons/md";
import NoteComposer from "./NoteComposer";

type Props = {
    isEdit: boolean;
    setIsEdit: React.Dispatch<React.SetStateAction<boolean>>;
    notes: {
        title: string | null | undefined;
        noteContent: string | null | undefined;
        isPinned: boolean;
        isArchived: boolean;
        isTrash: boolean;
    }[];
    addNote: (
        title: string | null | undefined,
        noteContent: string | null | undefined,
        isPinned: boolean,
        isArchived: boolean,
        isTrash: boolean
    ) => void;
    moveToTrash: (index: number) => void;
    // deleteNote: (index: number) => void;
};

const Notes = ({ isEdit, setIsEdit, notes, addNote, moveToTrash }: Props) => {
    return (
        <div>
            <NoteComposer
                isEdit={isEdit}
                setIsEdit={setIsEdit}
                addNote={addNote}
            />
            <div className="flex flex-wrap">
                {notes.map(
                    (note, index) =>
                        !note.isTrash && (
                            <div
                                key={index}
                                className="m-2 h-max w-60 border-2 border-solid"
                            >
                                <div className="whitespace-pre-wrap break-all p-3">
                                    <div className="btn-sm btn-circle btn float-right hover:bg-slate-200">
                                        <MdOutlinePushPin className="text-lg" />
                                    </div>
                                    <div className="font-semibold">
                                        {note.title}
                                    </div>
                                    <div>{note.noteContent}</div>
                                </div>
                                <div className="flex justify-evenly">
                                    <button className="btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black ">
                                        <MdOutlineNotificationAdd className="text-base" />
                                    </button>
                                    <button
                                        className="btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black disabled:cursor-not-allowed disabled:bg-inherit"
                                        disabled
                                    >
                                        <MdOutlinePersonAddAlt className="text-base" />
                                    </button>
                                    <button className="btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black ">
                                        <MdOutlinePalette className="text-base" />
                                    </button>
                                    <button className="btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black ">
                                        <MdOutlineImage className="text-base" />
                                    </button>
                                    <button className="btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black ">
                                        <MdOutlineArchive className="text-base" />
                                    </button>
                                    <div className="dropdown-end dropdown">
                                        <label
                                            tabIndex={0}
                                            className=" btn-sm btn-circle btn border-none bg-inherit text-slate-500 hover:bg-slate-200 hover:text-black "
                                        >
                                            <MdOutlineMoreVert className="text-base" />
                                        </label>
                                        <ul
                                            tabIndex={0}
                                            className="dropdown-content menu w-max bg-base-100 py-2 shadow-lg"
                                        >
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button
                                                    onClick={() =>
                                                        moveToTrash(index)
                                                    }
                                                    className="text-sm"
                                                >
                                                    Delete note
                                                </button>
                                            </li>
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button className="text-sm">
                                                    Add label
                                                </button>
                                            </li>
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button className="text-sm">
                                                    Add drawing
                                                </button>
                                            </li>
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button className="text-sm">
                                                    Make a copy
                                                </button>
                                            </li>
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button className="text-sm">
                                                    Show checkboxes
                                                </button>
                                            </li>
                                            <li className="py-1 px-4 hover:bg-gray-200">
                                                <button className="text-sm">
                                                    Copy to Google Docs
                                                </button>
                                            </li>
                                        </ul>
                                    </div>
                                </div>
                            </div>
                        )
                )}
            </div>
        </div>
    );
};

export default Notes;
