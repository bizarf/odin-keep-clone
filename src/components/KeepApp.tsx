import { React, useState, useEffect } from "react";
import { Routes, Route, useNavigate } from "react-router-dom";
import { onAuthStateChanged, getAuth, GoogleAuthProvider } from "firebase/auth";
import {
    getFirestore,
    doc,
    setDoc,
    getDoc,
    collection,
    onSnapshot,
} from "firebase/firestore";
import { app } from "./firebaseSetup";
import Header from "./Header";
import Nav from "./Nav";
import Notes from "./Notes";
import Reminders from "./Reminders";
import Archive from "./Archive";
import Trash from "./Trash";

type Props = {
    user: {
        accessToken: string;
        auth: object;
        displayName: string;
        email: string;
        emailVerified: boolean;
        isAnonymous: boolean;
        metadata: object;
        phoneNumber: unknown;
        photoURL: string;
        proactiveRefresh: object;
        providerData: Array<object>;
        providerId: string;
        reloadListener: unknown;
        reloadUserInfo: object;
        stsTokenManager: object;
        tenantId: unknown;
        uid: string;
        refreshToken: string;
    } | null;
    setUser: React.Dispatch<React.SetStateAction<object>>;
};

const KeepApp = ({ user, setUser }: Props) => {
    const auth = getAuth(app);
    const db = getFirestore(app);
    const navigate = useNavigate();

    const [notes, setNotes] = useState<
        Array<{
            title: string | null | undefined;
            noteContent: string | null | undefined;
            isPinned: boolean;
            isArchived: boolean;
            isTrash: boolean;
        }>
    >([]);

    const addNote = (
        title: string | null | undefined,
        noteContent: string | null | undefined,
        isPinned: boolean,
        isArchived: boolean,
        isTrash: boolean
    ) => {
        const newNote = {
            title: title,
            noteContent: noteContent,
            isPinned: isPinned,
            isArchived: isArchived,
            isTrash: isTrash,
        };
        if (title != "" && noteContent != "") {
            setNotes([...notes, newNote]);
        }

        if (
            (title === "" && noteContent != "") ||
            (title != "" && noteContent === "")
        ) {
            setNotes([...notes, newNote]);
        }
    };

    const moveToTrash = (index: number) => {
        const updatedNotes = notes.map((note, i) => {
            if (i === index) {
                note.isTrash = true;
                return note;
            } else {
                return note;
            }
        });
        setNotes([...updatedNotes]);
    };

    const [gridView, setGridView] = useState(true);

    const [isEdit, setIsEdit] = useState(false);

    // firebase stuff
    useEffect(() => {
        // if user is logged in, then send the data to the user state. if not then send them back to the splash page.
        onAuthStateChanged(auth, (googleUser) => {
            if (googleUser) {
                setUser(googleUser);
                // fetchUserData();
            } else {
                navigate("../");
            }
        });
    }, []);

    const fetchUserData = async () => {
        if (user) {
            const docRef = doc(db, "keep-data", `${user?.uid}`);
            const docSnap = await getDoc(docRef);

            if (docSnap.exists()) {
                // console.log("Document data:", docSnap.data());
                // console.log(docSnap.data().notes);
                setNotes([...docSnap.data().notes]);
            } else {
                // doc.data() will be undefined in this case
                console.log("No such document!");
                // create a new document that uses the user's google uid
                await setDoc(doc(collection(db, "keep-data"), `${user?.uid}`), {
                    notes: notes,
                });
            }
        }
    };

    useEffect(() => {
        console.log(user);
        if (user?.uid != "demo") {
            fetchUserData();
        }
    }, [user]);

    // whenever the notes array is update, we'll write to firestore
    useEffect(() => {
        const saveDataToFirestore = async () => {
            await setDoc(doc(collection(db, "keep-data"), `${user?.uid}`), {
                notes: notes,
            });
        };

        // function will only work if the user is logged in
        if (user) {
            if (user.uid != "demo") {
                saveDataToFirestore().catch(console.error);
            }
        }
    }, [notes]);

    return (
        <div className="grid h-screen grid-cols-[auto_1fr] grid-rows-[auto_1fr]">
            <Header gridView={gridView} setGridView={setGridView} user={user} />
            <Nav />
            <Routes>
                <Route
                    index
                    element={
                        <Notes
                            isEdit={isEdit}
                            setIsEdit={setIsEdit}
                            notes={notes}
                            addNote={addNote}
                            moveToTrash={moveToTrash}
                        />
                    }
                />
                <Route path="/reminders" element={<Reminders />} />
                <Route path="/archive" element={<Archive />} />
                <Route
                    path="/trash"
                    element={<Trash notes={notes} setNotes={setNotes} />}
                />
            </Routes>
        </div>
    );
};

export default KeepApp;
