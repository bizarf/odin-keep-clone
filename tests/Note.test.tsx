import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { describe, expect, it } from "vitest";
import { BrowserRouter } from "react-router-dom";
import KeepApp from "../src/components/KeepApp";
import Notes from "../src/components/Notes";

const user = {
    accessToken: "1",
    auth: { object: "object" },
    displayName: "Test",
    email: "test@test.com",
    emailVerified: true,
    isAnonymous: false,
    metadata: { object: "object" },
    phoneNumber: "",
    photoURL: "",
    proactiveRefresh: { object: "object" },
    providerData: [{ object: "object" }],
    providerId: "1",
    reloadListener: "",
    reloadUserInfo: { object: "object" },
    stsTokenManager: { object: "object" },
    tenantId: "1",
    uid: "demo",
    refreshToken: "1",
};

describe("note adding", () => {
    it("check if note composer has rendered on the note page", () => {
        render(<KeepApp user={user} setUser={undefined} />, {
            wrapper: BrowserRouter,
        });
        const composerPlaceholder = screen.getByText("Take a note...");
        expect(composerPlaceholder).toBeInTheDocument();
    });

    it("user successfully adds a note", async () => {
        render(<KeepApp user={user} setUser={undefined} />, {
            wrapper: BrowserRouter,
        });
        const notesTab = screen.getByText("Notes");
        await userEvent.click(notesTab);
        const composerPlaceholder = screen.getByText("Take a note...");
        await userEvent.click(composerPlaceholder);
        const textarea = screen.getByPlaceholderText("Take a note...");
        await userEvent.type(textarea, "This is a test note");
        await userEvent.click(screen.getByText("Close"));
        expect(screen.getByText("This is a test note")).toBeInTheDocument();
    });

    it("if the title and textcontent are empty, then the note won't be added", async () => {
        render(<KeepApp user={user} setUser={undefined} />, {
            wrapper: BrowserRouter,
        });
        const composerPlaceholder = screen.getByText("Take a note...");
        await userEvent.click(composerPlaceholder);
        await userEvent.click(screen.getByText("Close"));
        // expect(<KeepApp user={user} setUser={undefined} />).toMatchSnapshot();
    });

    it("the note will be posted if the title has text, but the textcontent is empty", async () => {
        render(<KeepApp user={user} setUser={undefined} />, {
            wrapper: BrowserRouter,
        });
        const composerPlaceholder = screen.getByText("Take a note...");
        await userEvent.click(composerPlaceholder);
        const titlePlaceholder = screen.getByPlaceholderText("Title");
        await userEvent.click(titlePlaceholder);
        await userEvent.type(titlePlaceholder, "This is a test title");
        await userEvent.click(screen.getByText("Close"));
        expect(screen.getByText("This is a test title")).toBeInTheDocument();
    });

    it("the note will be posted if the textcontent has text, but the title is empty", async () => {
        render(<KeepApp user={user} setUser={undefined} />, {
            wrapper: BrowserRouter,
        });
        const composerPlaceholder = screen.getByText("Take a note...");
        await userEvent.click(composerPlaceholder);
        const textarea = screen.getByPlaceholderText("Take a note...");
        await userEvent.type(textarea, "This is a test note");
        await userEvent.click(screen.getByText("Close"));
        expect(screen.getByText("This is a test note")).toBeInTheDocument();
    });
});
