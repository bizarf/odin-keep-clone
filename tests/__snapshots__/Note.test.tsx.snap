// Vitest Snapshot v1

exports[`note adding > if the title and textcontent are empty, then the note won't be added 1`] = `<Notes />`;
